File: app/src/test/java/com/coinninja/coinkeeper/ui/settings/SettingsActivityTest.java
package com.coinninja.coinkeeper.ui.settings;

import android.app.DialogFragment;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.Switch;

import com.coinninja.coinkeeper.R;
import com.coinninja.coinkeeper.TestCoinKeeperApplication;
import com.coinninja.coinkeeper.cn.wallet.CNWalletManager;
import com.coinninja.coinkeeper.cn.wallet.SyncWalletManager;
import com.coinninja.coinkeeper.cn.wallet.dust.DustProtectionPreference;
import com.coinninja.coinkeeper.ui.backup.BackupRecoveryWordsStartActivity;
import com.coinninja.coinkeeper.util.Intents;
import com.coinninja.coinkeeper.util.PhoneNumberUtil;
import com.coinninja.coinkeeper.util.uri.DropbitUriBuilder;
import com.coinninja.coinkeeper.view.activity.AuthorizedActionActivity;
import com.coinninja.coinkeeper.view.activity.LicensesActivity;
import com.coinninja.coinkeeper.view.activity.SplashActivity;
import com.coinninja.matchers.IntentMatcher;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.android.controller.ActivityController;
import org.robolectric.shadows.ShadowActivity;

import static com.coinninja.android.helpers.Views.clickOn;
import static com.coinninja.android.helpers.Views.withId;
import static junit.framework.TestCase.assertNotNull;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.robolectric.Shadows.shadowOf;

@RunWith(RobolectricTestRunner.class)
public class SettingsActivityTest {

    @Mock
    DustProtectionPreference dustProtectionPreference;
    @Mock
    PhoneNumberUtil phoneNumberUtil;
    @Mock
    private CNWalletManager cnWalletManager;
    @Mock
    private DeleteWalletPresenter deleteWalletPresenter;
    @Mock
    private SyncWalletManager syncWalletManager;
    private DropbitUriBuilder dropbitUriBuilder = new DropbitUriBuilder();
    private ActivityController<SettingsActivity> activityController;
    private SettingsActivity activity;
    private ShadowActivity shadowActivity;

    @After
    public void tearDown() {
        cnWalletManager = null;
        deleteWalletPresenter = null;
        syncWalletManager = null;
        activity = null;
        activityController = null;
        shadowActivity = null;
        phoneNumberUtil = null;
        dropbitUriBuilder = null;
        dustProtectionPreference = null;
    }

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        TestCoinKeeperApplication coinKeeperApplication = (TestCoinKeeperApplication) RuntimeEnvironment.application;
        when(coinKeeperApplication.authentication.isAuthenticated()).thenReturn(true);
        when(dustProtectionPreference.isDustProtectionEnabled()).thenReturn(false);
        activityController = Robolectric.buildActivity(SettingsActivity.class);
        activityController.create();
        activity = activityController.get();
        activity.deleteWalletPresenter = deleteWalletPresenter;
        activity.cnWalletManager = cnWalletManager;
        activity.syncWalletManager = syncWalletManager;
        activity.phoneNumberUtil = phoneNumberUtil;
        activity.dustProtectionPreference = dustProtectionPreference;
        activity.dropbitUriBuilder = dropbitUriBuilder;
        shadowActivity = shadowOf(activity);
    }

    @Test
    public void shows_not_backed_up_when_user_skipped_backup() {
        when(cnWalletManager.hasSkippedBackup()).thenReturn(true);
        start();

        assertThat(activity.findViewById(R.id.not_backed_up_message).getVisibility(),
                equalTo(View.VISIBLE));

    }

    @Test
    public void hides_not_backed_up_when_words_backedUp() {
        when(cnWalletManager.hasSkippedBackup()).thenReturn(false);
        start();

        assertThat(activity.findViewById(R.id.not_backed_up_message).getVisibility(),
                equalTo(View.GONE));
    }

    @Test
    public void shows_recovery_words_help_screen() {
        start();

        activity.findViewById(R.id.recover_wallet).performClick();

        Intent startedActivity = shadowActivity.getNextStartedActivity();
        assertThat(startedActivity.getComponent().getClassName(),
                equalTo(BackupRecoveryWordsStartActivity.class.getName()));
    }

    @Test
    public void shows_licenses() {
        start();

        activity.findViewById(R.id.open_source).performClick();

        Intent startedActivity = shadowActivity.getNextStartedActivity();
        assertThat(startedActivity.getComponent().getClassName(),
                equalTo(LicensesActivity.class.getName()));
    }

    @Test
    public void clicking_sync_triggers_sync() {
        start(true);

        activity.findViewById(R.id.settings_sync).performClick();

        verify(syncWalletManager).syncNow();
    }

    @Test
    public void shows_sync_when_debug_is_on() {
        start(true);

        assertThat(activity.findViewById(R.id.settings_sync).getVisibility(), equalTo(View.VISIBLE));
    }

    @Test
    public void hides_sync_when_debug_is_off() {
        start();

        assertThat(activity.findViewById(R.id.settings_sync).getVisibility(), equalTo(View.GONE));
    }

    @Test
    public void navigates_to_splash_activity_on_Delete() {
        start();

        activity.onDeleted();

        Intent startedActivity = shadowActivity.getNextStartedActivity();
        assertThat(startedActivity.getComponent().getClassName(), equalTo(SplashActivity.class.getName()));
        assertThat(startedActivity.getFlags(), equalTo(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK));
    }

    @Test
    public void performs_authorization_for_deleting_wallet() {
        start();
        activity.findViewById(R.id.delete_wallet).performClick();
        DialogFragment dialog = (DialogFragment) activity.getFragmentManager().findFragmentByTag(SettingsActivity.TAG_CONFIRM_DELETE_WALLET);

        activity.onClick(dialog.getDialog(), DialogInterface.BUTTON_POSITIVE);

        ShadowActivity.IntentForResult activityForResult = shadowActivity.getNextStartedActivityForResult();
        assertThat(activityForResult.requestCode, equalTo(SettingsActivity.DELETE_WALLET_REQUEST_CODE));
        assertThat(activityForResult.intent.getComponent().getClassName(), equalTo(AuthorizedActionActivity.class.getName()));
    }

    @Test
    public void performs_delete() {
        start();

        activity.onActivityResult(SettingsActivity.DELETE_WALLET_REQUEST_CODE, AuthorizedActionActivity.RESULT_AUTHORIZED, null);

        verify(deleteWalletPresenter).onDelete();
    }

    @Test
    public void negative_delete_confirmation_dismisses_only() {
        start();
        activity.findViewById(R.id.delete_wallet).performClick();
        DialogFragment dialog = (DialogFragment) activity.getFragmentManager().findFragmentByTag(SettingsActivity.TAG_CONFIRM_DELETE_WALLET);

        activity.onClick(dialog.getDialog(), DialogInterface.BUTTON_NEGATIVE);

        assertNull(activity.getFragmentManager().findFragmentByTag(SettingsActivity.TAG_CONFIRM_DELETE_WALLET));
    }

    @Test
    public void prompts_to_confirm_delete() {
        start();

        activity.findViewById(R.id.delete_wallet).performClick();

        verify(deleteWalletPresenter).setCallback(any());
        assertNotNull(activity.getFragmentManager().findFragmentByTag(SettingsActivity.TAG_CONFIRM_DELETE_WALLET));
    }

    @Test
    public void show_auth_delete_with_message() {
        String expectedAuthMessage = "Enter pin to confirm deletion of your wallet.";
        start();

        activity.authorizeDelete();
        Intent intent = shadowActivity.getNextStartedActivity();

        Bundle extras = intent.getExtras();
        String authorizedActionMessage = extras.getString(Intents.EXTRA_AUTHORIZED_ACTION_MESSAGE);

        assertThat(authorizedActionMessage, equalTo(expectedAuthMessage));
    }

    @Test
    public void renders_state_of_dust_protection_preference() {
        when(dustProtectionPreference.isDustProtectionEnabled()).thenReturn(true);
        start();

        Switch view = withId(activity, R.id.dust_protection_toggle);

        assertTrue(view.isChecked());
    }

    @Test
    public void observes_switch_toggle() {
        when(dustProtectionPreference.isDustProtectionEnabled()).thenReturn(true);
        start();
        clickOn(withId(activity, R.id.dust_protection_toggle));

        verify(dustProtectionPreference).setProtection(false);
    }

    @Test
    public void clicking_on_tooltip_for_dust_protection_navigates_to_website() {
        start();

        clickOn(withId(activity, R.id.dust_protection_tooltip));

        Intent intent = shadowOf(activity).getNextStartedActivity();
        Intent expectedIntent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://dropbit.app/tooltips/dustprotection"));
        assertThat(intent, IntentMatcher.equalTo(expectedIntent));
    }

    private void start() {
        start(false);
    }

    private void start(boolean isDebug) {
        activity.isDebugBuild = isDebug;
        activityController.start().resume().visible();
    }
}

File: app/src/test/java/com/coinninja/coinkeeper/ui/settings/SettingsActivityTest.kt
package com.coinninja.coinkeeper.ui.settings

import android.content.DialogInterface
import android.content.Intent
import android.view.View
import android.widget.Switch
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.MutableLiveData
import androidx.test.core.app.ActivityScenario
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.coinninja.android.helpers.Views.clickOn
import com.coinninja.android.helpers.Views.withId
import com.coinninja.coinkeeper.R
import com.coinninja.coinkeeper.TestCoinKeeperApplication
import com.coinninja.coinkeeper.cn.service.YearlyHighViewModel
import com.coinninja.coinkeeper.cn.wallet.dust.DustProtectionPreference
import com.coinninja.coinkeeper.ui.backup.BackupRecoveryWordsStartActivity
import com.coinninja.coinkeeper.util.DropbitIntents
import com.coinninja.coinkeeper.util.analytics.Analytics
import com.coinninja.coinkeeper.view.activity.AuthorizedActionActivity
import com.coinninja.coinkeeper.view.activity.BackupActivity
import com.coinninja.coinkeeper.view.activity.LicensesActivity
import com.coinninja.coinkeeper.view.activity.SplashActivity
import com.coinninja.coinkeeper.view.dialog.GenericAlertDialog
import com.google.common.truth.Truth.assertThat
import com.nhaarman.mockitokotlin2.whenever
import junit.framework.TestCase.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.robolectric.Shadows.shadowOf

@Suppress("UNCHECKED_CAST")
@RunWith(AndroidJUnit4::class)
class SettingsActivityTest {

    @Before
    fun configureDI() {
        val coinKeeperApplication = ApplicationProvider.getApplicationContext<TestCoinKeeperApplication>()
        coinKeeperApplication.dustProtectionPreference = mock(DustProtectionPreference::class.java)
        coinKeeperApplication.yearlyHighViewModel = mock(YearlyHighViewModel::class.java)
        val liveData: MutableLiveData<Boolean> = mock(MutableLiveData::class.java) as MutableLiveData<Boolean>
        whenever(coinKeeperApplication.yearlyHighViewModel.isSubscribedToYearlyHigh).thenReturn(liveData)
        whenever(coinKeeperApplication.authentication.isAuthenticated).thenReturn(true)
    }

    fun createScenario(): ActivityScenario<SettingsActivity> {
        val scenario = ActivityScenario.launch(SettingsActivity::class.java)
        scenario.onActivity { activity ->
            whenever(activity.dustProtectionPreference.isDustProtectionEnabled).thenReturn(false)
        }
        return scenario
    }

    @Test
    fun shows_not_backed_up_when_user_skipped_backup() {
        val coinKeeperApplication = ApplicationProvider.getApplicationContext<TestCoinKeeperApplication>()
        whenever(coinKeeperApplication.cnWalletManager.hasSkippedBackup()).thenReturn(true)
        val scenario = createScenario()
        scenario.moveToState(Lifecycle.State.RESUMED)

        scenario.onActivity { activity ->
            assertThat(activity.findViewById<View>(R.id.not_backed_up_message).visibility).isEqualTo(View.VISIBLE)
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun hides_not_backed_up_when_words_backedUp() {
        val scenario = createScenario()
        scenario.onActivity { activity ->
            whenever(activity.cnWalletManager.hasSkippedBackup()).thenReturn(false)
        }
        scenario.moveToState(Lifecycle.State.RESUMED)

        scenario.onActivity { activity ->
            assertThat(activity.findViewById<View>(R.id.not_backed_up_message).visibility).isEqualTo(View.GONE)
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun shows_recovery_words_help_screen() {
        val scenario = createScenario()
        scenario.moveToState(Lifecycle.State.RESUMED)

        scenario.onActivity { activity ->
            activity.findViewById<View>(R.id.recover_wallet).performClick()
            val shadowActivity = shadowOf(activity)
            val startedActivity = shadowActivity.nextStartedActivity
            assertThat(startedActivity.component!!.className).isEqualTo(BackupRecoveryWordsStartActivity::class.java.name)
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun shows_licenses() {
        val scenario = createScenario()
        scenario.moveToState(Lifecycle.State.RESUMED)

        scenario.onActivity { activity ->
            activity.findViewById<View>(R.id.open_source).performClick()
            val shadowActivity = shadowOf(activity)
            val startedActivity = shadowActivity.nextStartedActivity
            assertThat(startedActivity.component!!.className).isEqualTo(LicensesActivity::class.java.name)
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun navigates_to_splash_activity_on_Delete() {
        val scenario = createScenario()
        scenario.moveToState(Lifecycle.State.RESUMED)

        scenario.onActivity { activity ->
            activity.onDeleted()
            val shadowActivity = shadowOf(activity)
            val startedActivity = shadowActivity.nextStartedActivity
            assertThat(startedActivity.component!!.className).isEqualTo(SplashActivity::class.java.name)
            assertThat(startedActivity.flags).isEqualTo(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun performs_authorization_for_deleting_wallet() {
        val scenario = createScenario()

        scenario.moveToState(Lifecycle.State.RESUMED)

        scenario.onActivity { activity ->
            activity.findViewById<View>(R.id.delete_wallet).performClick()
            val dialog = activity.supportFragmentManager.findFragmentByTag(SettingsActivity.TAG_CONFIRM_DELETE_WALLET) as GenericAlertDialog
            val shadowActivity = shadowOf(activity)

            activity.onClick(dialog.alertDialog, DialogInterface.BUTTON_POSITIVE)

            val activityForResult = shadowActivity.nextStartedActivityForResult
            assertThat(activityForResult.requestCode).isEqualTo(SettingsActivity.DELETE_WALLET_REQUEST_CODE)
            assertThat(activityForResult.intent.component!!.className).isEqualTo(AuthorizedActionActivity::class.java.name)
            activity.onDeleted()
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun performs_delete() {
        val scenario = createScenario()

        scenario.moveToState(Lifecycle.State.RESUMED)

        scenario.onActivity { activity ->
            activity.onActivityResult(SettingsActivity.DELETE_WALLET_REQUEST_CODE, AuthorizedActionActivity.RESULT_AUTHORIZED, null)

            verify(activity.deleteWalletPresenter).onDelete()
        }
    }

    @Test
    fun negative_delete_confirmation_dismisses_only() {
        val scenario = createScenario()

        scenario.moveToState(Lifecycle.State.RESUMED)

        scenario.onActivity { activity ->
            activity.findViewById<View>(R.id.delete_wallet).performClick()
            val dialog = activity.supportFragmentManager.findFragmentByTag(SettingsActivity.TAG_CONFIRM_DELETE_WALLET) as GenericAlertDialog

            activity.onClick(dialog.alertDialog, DialogInterface.BUTTON_NEGATIVE)

            assertNull(activity.supportFragmentManager.findFragmentByTag(SettingsActivity.TAG_CONFIRM_DELETE_WALLET))
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun prompts_to_confirm_delete() {
        val scenario = createScenario()

        scenario.moveToState(Lifecycle.State.RESUMED)

        scenario.onActivity { activity ->
            activity.findViewById<View>(R.id.delete_wallet).performClick()

            verify(activity.deleteWalletPresenter).setCallback(ArgumentMatchers.any())
            assertNotNull(activity.supportFragmentManager.findFragmentByTag(SettingsActivity.TAG_CONFIRM_DELETE_WALLET))
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun show_auth_delete_with_message() {
        val expectedAuthMessage = "Enter pin to confirm deletion of your wallet."
        val scenario = createScenario()

        scenario.moveToState(Lifecycle.State.RESUMED)
        scenario.onActivity { activity ->
            activity.authorizeDelete()

            val intent = shadowOf(activity).nextStartedActivity
            val extras = intent.extras
            val authorizedActionMessage = extras!!.getString(DropbitIntents.EXTRA_AUTHORIZED_ACTION_MESSAGE)
            assertThat(authorizedActionMessage).isEqualTo(expectedAuthMessage)
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun renders_state_of_dust_protection_preference() {
        whenever(ApplicationProvider.getApplicationContext<TestCoinKeeperApplication>().dustProtectionPreference.isDustProtectionEnabled).thenReturn(true)
        val scenario = createScenario()

        scenario.moveToState(Lifecycle.State.RESUMED)
        scenario.onActivity { activity ->
            val view = withId<Switch>(activity, R.id.dust_protection_toggle)
            assertTrue(view!!.isChecked)
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun observes_switch_toggle_dust_protection() {
        whenever(ApplicationProvider.getApplicationContext<TestCoinKeeperApplication>().dustProtectionPreference.isDustProtectionEnabled).thenReturn(true)
        val scenario = createScenario()

        scenario.moveToState(Lifecycle.State.RESUMED)
        scenario.onActivity { activity ->
            clickOn(withId(activity, R.id.dust_protection_toggle))

            verify(activity.dustProtectionPreference).setProtection(false)
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun clicking_on_tooltip_for_dust_protection_navigates_to_website() {
        val scenario = createScenario()
        scenario.onActivity { activity ->
            whenever(activity.dustProtectionPreference.isDustProtectionEnabled).thenReturn(true)
        }

        scenario.moveToState(Lifecycle.State.RESUMED)
        scenario.onActivity { activity ->
            clickOn(withId(activity, R.id.dust_protection_tooltip))

            val intent = shadowOf(activity).nextStartedActivity
            assertThat(intent.action).isEqualTo(Intent.ACTION_VIEW)
            assertThat(intent.data.toString()).isEqualTo("https://dropbit.app/tooltips/dustprotection")
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun clicking_yearly_high_subscription_toggles_subscription() {
        val scenario = createScenario()

        scenario.moveToState(Lifecycle.State.RESUMED)
        scenario.onActivity { activity ->
            activity.yearlyHighObserver.onChanged(true)
            clickOn(activity.findViewById<Switch>(R.id.yearly_high_subscription))
            verify(activity.yearlyHighViewModel).toggleSubscription(true)

            scenario.recreate()

            activity.yearlyHighObserver.onChanged(false)
            clickOn(activity.findViewById<Switch>(R.id.yearly_high_subscription))
            verify(activity.yearlyHighViewModel).toggleSubscription(false)
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun loads_yearly_high_subscription_state_when_resumed() {
        val scenario = createScenario()

        scenario.moveToState(Lifecycle.State.RESUMED)
        scenario.onActivity { activity ->
            verify(activity.yearlyHighViewModel.isSubscribedToYearlyHigh).observe(activity, activity.yearlyHighObserver)
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun updates_yearly_high_when_subscription_changes() {
        val scenario = createScenario()

        scenario.moveToState(Lifecycle.State.RESUMED)
        scenario.onActivity { activity ->
            activity.yearlyHighObserver.onChanged(true)

            assertThat(withId<Switch>(activity, R.id.yearly_high_subscription)!!.isChecked).isTrue()
        }
        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun shows_legacy_words_when_user_has_segwit_and_legacy() {
        val scenario = createScenario()
        scenario.onActivity { activity ->
            // Gone when no legacy wallet
            assertThat(activity.legacyWalletButton.visibility).isEqualTo(View.GONE)
            assertThat(activity.legacyWalletDivider.visibility).isEqualTo(View.GONE)

            // Visible when legacy wallet exists
            whenever(activity.cnWalletManager.hasLegacyWallet).thenReturn(true)
            activity.setupLegacyWallet()

            assertThat(activity.legacyWalletButton.visibility).isEqualTo(View.VISIBLE)
            assertThat(activity.legacyWalletDivider.visibility).isEqualTo(View.VISIBLE)
        }

        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun clicking_on_legacy_wallet_shows_authroizes_action() {
        val scenario = createScenario()
        scenario.onActivity { activity ->
            whenever(activity.cnWalletManager.hasLegacyWallet).thenReturn(true)
            activity.setupLegacyWallet()

            activity.legacyWalletButton.performClick()
            val intentForResult = shadowOf(activity).peekNextStartedActivityForResult()

            assertThat(intentForResult).isNotNull()
            assertThat(intentForResult.requestCode).isEqualTo(SettingsActivity.legacyWalletRequestCode)
            assertThat(intentForResult.intent.component.className).isEqualTo(AuthorizedActionActivity::class.java.name)
        }

        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }

    @Test
    fun shows_legacy_words_once_authorized() {
        val scenario = createScenario()
        scenario.onActivity { activity ->
            val words: Array<String> = arrayOf("word", "word", "word", "word", "word", "word", "word", "word", "word", "word", "word", "word")
            whenever(activity.cnWalletManager.legacyWords).thenReturn(words)

            activity.onActivityResult(SettingsActivity.legacyWalletRequestCode, AuthorizedActionActivity.RESULT_AUTHORIZED, null)

            val intent = shadowOf(activity).peekNextStartedActivity()
            assertThat(intent.component.className).isEqualTo(BackupActivity::class.java.name)
            assertThat(intent.getStringArrayExtra(DropbitIntents.EXTRA_RECOVERY_WORDS)).isEqualTo(words)
            assertThat(intent.getIntExtra(DropbitIntents.EXTRA_VIEW_STATE, -1)).isEqualTo(DropbitIntents.EXTRA_VIEW)
            verify(activity.analytics).trackEvent(Analytics.EVENT_VIEW_LEGACY_WORDS)
        }

        scenario.moveToState(Lifecycle.State.DESTROYED)
        scenario.close()
    }
}